File: app/src/main/java/com/simcom/ecashier/model/room/PersonToGroupDao.java
package com.simcom.ecashier.model.room;

import androidx.lifecycle.LiveData;
import androidx.room.Dao;
import androidx.room.Delete;
import androidx.room.Insert;
import androidx.room.Query;

import java.util.List;

@Dao
public interface PersonToGroupDao {
    @Insert
    void insert(PersonToGroup item);

    @Delete
    void delete(PersonToGroup item);

    @Query("SELECT p.*, cl.collectionId, (SELECT count(*) FROM COLLECTION_LOG_TABLE ORDER BY time) AS rank " +
            "FROM PERSON_TO_GROUP_TABLE AS ptg " +
            "INNER JOIN PERSON_TABLE AS p ON ptg.personId = p.id " +
            "LEFT OUTER JOIN COLLECTION_LOG_TABLE AS cl ON p.id = cl.personId & cl.collectionId = :collectionId " +
            "WHERE ptg.groupId = :groupId " +
            "ORDER BY p.name ")
    LiveData<List<PersonExtended>> getPeopleByCollection(int collectionId, int groupId);

    @Query("SELECT COUNT(*) FROM PERSON_TO_GROUP_TABLE WHERE groupId = :groupId")
    int getGroupSize(int groupId);
}

File: app/src/main/java/com/simcom/ecashier/model/room/PersonToGroupDao.kt
package com.simcom.ecashier.model.room

import androidx.lifecycle.LiveData
import androidx.room.*

@Dao
interface PersonToGroupDao {
    @Insert
    suspend fun insert(item: PersonToGroup)

    @Delete
    suspend fun delete(item: PersonToGroup)

    @Query("DELETE FROM person_to_group_table WHERE personId = :personId AND groupId = :groupId")
    suspend fun deleteWithIds(personId: Long, groupId: Long)

    @Query("DELETE FROM person_to_group_table WHERE groupId = :groupId")
    suspend fun deleteWithGroupId(groupId: Long)

    @Query("SELECT PERSON_TABLE.* " +
            "FROM person_to_group_table " +
            "INNER JOIN PERSON_TABLE ON PERSON_TABLE.id = PERSON_TO_GROUP_TABLE.personId " +
            "WHERE groupId = :groupId")
    fun getPeopleFromGroup(groupId: Long) : LiveData<List<Person>>

    @Query("SELECT PERSON_TABLE.* " +
            "FROM person_to_group_table " +
            "INNER JOIN PERSON_TABLE ON PERSON_TABLE.id = PERSON_TO_GROUP_TABLE.personId " +
            "WHERE groupId = :groupId")
    suspend fun getPeopleFromGroupSynchronous(groupId: Long) : List<Person>

    @Query("SELECT COUNT(*) FROM PERSON_TO_GROUP_TABLE WHERE groupId = :groupId")
    fun getGroupSize(groupId: Long): Int
}